packages/core/src/browser/widget-manager.ts

```diff
@@ -222,6 +218,34 @@ export class WidgetManager {
         }
     }
 
+    /**
+     * Creates a new widget or returns the existing widget for the given description.
+     * @param factoryId the widget factory id.
+     * @param options the widget factory specific information.
+     *
+     * @returns a promise resolving to the widget.
+     */
+    async getOrCreateWidget<T extends Widget>(factoryId: string, options?: any): Promise<T> {
+        const key = this.toKey({ factoryId, options });
+        const existingWidget = this.doGetWidget<T>(key);
+        if (existingWidget) {
+            return existingWidget;
+        }
+        return this.createWidget(factoryId, options);
+    }
+
+    /**
+     * Returns whether the wedget exists or not.
+     * @param factoryId The widget factory id.
+     * @param options The widget factory specific information.
+     *
+     * @returns Whether the wedget exists or not.
+     */
+    isWidgetExisting(factoryId: string, options?: any): boolean {
```

Comment:
The naming of the method can be improved.
We can simplify it to `widgetExists` or something along the lines of:

```suggestion
    findWidget(factoryId: string, options?: any): Widget | undefined {
```

If the widget exists, simply return it.

Reply:
I prefer to choose renaming the method.
Maybe the best way would change `tryGetWidget` to get a widget from `widgetPromises` and `pendingWidgetPromises` not only from `widgetPromises`, if possible.

Reply:
Changed the name.